src/promql/src/functions/extrapolate_rate.rs

```diff
@@ -0,0 +1,564 @@
+    /// Range duration in millisecond
+    range_length: i64,
+}
+
+impl<const IS_COUNTER: bool, const IS_RATE: bool> ExtrapolatedRate<IS_COUNTER, IS_RATE> {
+    /// Constructor. Other public usage should use [`scalar_udf`] instead.
+    fn new(range_length: i64) -> Self {
+        Self { range_length }
+    }
+
+    fn input_type() -> Vec<DataType> {
+        vec![
+            // timestamp range vector
+            RangeArray::convert_data_type(DataType::Timestamp(TimeUnit::Millisecond, None)),
+            // value range vector
+            RangeArray::convert_data_type(DataType::Float64),
+            // timestamp vector
+            DataType::Timestamp(TimeUnit::Millisecond, None),
+        ]
+    }
+
+    fn return_type() -> DataType {
+        DataType::Float64
+    }
+
+    fn calc(&self, input: &[ColumnarValue]) -> Result<ColumnarValue, DataFusionError> {
+        assert_eq!(input.len(), 3);
+
+        // construct matrix from input
+        let ts_array = extract_array(&input[0])?;
+        let ts_range = RangeArray::try_new(ts_array.data().clone().into())?;
+        let value_array = extract_array(&input[1])?;
+        let value_range = RangeArray::try_new(value_array.data().clone().into())?;
+        let ts = extract_array(&input[2])?;
+        let ts = ts
+            .as_any()
+            .downcast_ref::<TimestampMillisecondArray>()
+            .unwrap();
+
+        // calculation
+        let mut result_array = Vec::with_capacity(ts_range.len());
+        for index in 0..ts_range.len() {
+            let timestamps = ts_range.get(index).unwrap();
+            let timestamps = timestamps
+                .as_any()
+                .downcast_ref::<TimestampMillisecondArray>()
+                .unwrap()
+                .values();
+            let end_ts = ts.value(index);
+            let values = value_range.get(index).unwrap();
+            let values = values
+                .as_any()
+                .downcast_ref::<Float64Array>()
+                .unwrap()
+                .values();
+
+            if values.len() < 2 {
+                result_array.push(None);
+                continue;
+            }
+
+            // refer to functions.go L83-L110
+            let mut result_value = values.last().unwrap() - values.first().unwrap();
+            if IS_COUNTER {
+                for window in values.windows(2) {
+                    let prev = window[0];
+                    let curr = window[1];
+                    if curr < prev {
+                        result_value += prev
+                    }
+                }
+            }
+
+            let mut factor = Self::extrapolate_factor(
+                timestamps,
+                end_ts,
+                self.range_length,
+                *values.first().unwrap(),
+                result_value,
+            );
+
+            if IS_RATE {
+                factor /= self.range_length as f64 / 1000.0;
```

Comment:
Is it possible that the `self.range_length` is zero?

Reply:
This is checked during parsing. But I should also add an assertion here 👍